metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -129,155 +161,519 @@ func getResource(resourceName string) kubernetes.Resource {
+	if !ok {
+		// Check if we need to add namespace to the watcher's options.
+		if isNamespaced(resourceName) {
+			options.Namespace = namespace
+		}
+		watcher, err := kubernetes.NewNamedWatcher(resourceName, client, resource, options, nil)
+		if err != nil {
+			return false, err
+		}
+		resourceWatchers.metaWatchersMap[resourceName] = &metaWatcher{
+			watcher:         watcher,
+			started:         false, // not started yet
+			metadataObjects: make(map[string]bool),
+			enrichers:       make(map[string]*enricher),
+			metricsetsUsing: make([]string, 0),
+			restartWatcher:  nil,
+			nodeScope:       nodeScope,
+		}
+		return true, nil
+	} else if resourceMetaWatcher.nodeScope != nodeScope && resourceMetaWatcher.nodeScope {
+		// It might happen that the watcher already exists, but is only being used to monitor the resources
+		// of a single node(e.g. created by pod metricset). In that case, we need to check if we are trying to create a new watcher that will track
+		// the resources of whole cluster(e.g. in case of state_pod metricset).
+		// If it is the case, then we need to update the watcher by changing its watch options (removing options.Node)
+		// A running watcher cannot be updated directly. Instead, we must create a new one with the correct watch options.
+		// The new restartWatcher must be identical to the old watcher, including the same handler function, with the only difference being the watch options.
+
+		if isNamespaced(resourceName) {
+			options.Namespace = namespace
+		}
+		restartWatcher, err := kubernetes.NewNamedWatcher(resourceName, client, resource, options, nil)
+		if err != nil {
+			return false, err
+		}
+		// update the handler of the restartWatcher to match the current watcher's handler.
+		restartWatcher.AddEventHandler(resourceMetaWatcher.watcher.GetEventHandler())
+		resourceMetaWatcher.restartWatcher = restartWatcher
+		resourceMetaWatcher.nodeScope = nodeScope
+	}
+	return false, nil
+}
+
+// addToMetricsetsUsing adds metricset identified by metricsetUsing to the list of resources using the shared watcher
+// identified by resourceName.
+func addToMetricsetsUsing(resourceName string, metricsetUsing string, resourceWatchers *Watchers) {
+	resourceWatchers.lock.Lock()
+	defer resourceWatchers.lock.Unlock()
+
+	data, ok := resourceWatchers.metaWatchersMap[resourceName]
+	if ok {
+		contains := false
+		for _, which := range data.metricsetsUsing {
+			if which == metricsetUsing {
+				contains = true
+				break
+			}
+		}
+		// add this resource to the list of resources using it
+		if !contains {
+			data.metricsetsUsing = append(data.metricsetsUsing, metricsetUsing)
+		}
+	}
+}
+
+// removeFromMetricsetsUsing removes the metricset from the list of resources using the shared watcher.
+// It returns true if element was removed and new size of array.
+// The cache should be locked when called.
+func removeFromMetricsetsUsing(resourceName string, notUsingName string, resourceWatchers *Watchers) (bool, int) {
```

Comment:
The inverse of this function `addToMetricsetsUsing` grabs the lock and this function doesn't. That inconsistency is not clear. I would prefer to see that consistent between the two functions.